File: tagalert.py
##################################################
## TagAlertBot (https://telegram.me/tagalertbot) #
## Simple notifications for mentions             #
##                                               #
## Author: Antonio Pitasi (@Zaphodias)           #
## 2016 - made with love                         #
##################################################

import telebot
import shelve
from config import *

replies = {
    'start_group':      'Contact me in private for more infos and enabling me.',
    'start_private':
'Hello.\n\
You are now enabled to receive notifications from me, \
Just add @TagAlertBot in your groups and I\'ll start working.\n\
When you\'ll get tagged I\'ll send a message to you.\n\n\
Source code and infos: http://tagalert.pitasi.space/\n\
 - a bot by @Zaphodias.',
    'main_text':        '<b>[ Incoming Message ]</b>\n\n<b>[ FROM ]</b>\n' +
                        u'\U0001f464' +
                        '  {}\n<b>[ GROUP ]</b>\n' +
                        u'\U0001f465' +
                        '  {}\n<b>[ TEXT ]</b> \n' +
                        u'\u2709\ufe0f' + '  {}',
    'options':          'From group: <b>{}</b>\nAvailable operations:',
    'retrieve':         'Find the message',
    'retrieve_group':   'Here is your message, @{}.',
    'retrieve_success': 'Done!\nNow check the group of the message.',
    'no_username':      'Sorry.\nYou need to set an username from Telegram\'s settings before using me.',
    'error':            'Sorry.\nSomething went wrong.'
}
# end configuration

d   = shelve.open(config['db_path'])
bot = telebot.TeleBot(config['token'], threaded=False)
n = bot.get_me()
group_flood = {}

def remove_user(username):
    if not username: return
    try:
        print("Removing {} ({}) from database".format(username, d[username]))
        del d[username]
        d.sync()
    except KeyError: pass

def add_user(username, user_id):
    if not username or not user_id: return
    try:
        d[username.lower()] = user_id
        d.sync()
    except Exception as e: print(e)

@bot.callback_query_handler(func=lambda call: call.data[:9] == "/retrieve")
def callback_handler(call):
    m_id = call.data[10:].split('_')
    try:
        bot.send_message(-int(m_id[1]),
                        replies['retrieve_group'].format(call.from_user.username),
                        reply_to_message_id=int(m_id[0]))
        bot.answer_callback_query(call.id, text=replies['retrieve_success'], show_alert=True)
    except Exception as e:
        print("[CALLBACK HANDLER EXCEPTION] [{} - {}]\n{}".format(call.from_user.username, call.from_user.id, e))
        try: bot.answer_callback_query(call.id, text=replies['error'], show_alert=True)
        except Exception: pass

@bot.message_handler(commands=['help', 'start'])
def help_handler(m):
    if not m.from_user.username:
        try: bot.reply_to(m, replies['no_username'])
        except Exception: pass

        return

    add_user(m.from_user.username, m.from_user.id)
    is_group = m.chat.type == 'group' or m.chat.type == 'supergroup'
    try:
        last_message_time = group_flood[m.chat.id] if m.chat.id in group_flood else None

        if (not last_message_time or m.date > last_message_time + config['help_flood_time']):
            bot.reply_to(m, replies['start_group'] if is_group else replies['start_private'])
            group_flood[m.chat.id] = m.date

    except telebot.apihelper.ApiException as e:
        if e.result.status_code == 403 or e.result.status_code == 400:
            remove_user(m.from_user.username)
    except Exception as e:
        print("[HELP HANDLER EXCEPTION] [{} - {}]\n{}".format(m.from_user.username, m.from_user.id, e))

@bot.message_handler(func=lambda m:                           \
                                m.entities and                  \
                                (m.chat.type == 'group' or      \
                                 m.chat.type == 'supergroup')   )
def main_handler(m):
    markup = telebot.types.InlineKeyboardMarkup()
    if m.chat.username:
        btn = telebot.types.InlineKeyboardButton(replies['retrieve'], url="telegram.me/{}/{}".format(m.chat.username, m.message_id))
    else:
        btn = telebot.types.InlineKeyboardButton(replies['retrieve'], callback_data="/retrieve_{}_{}".format(m.message_id, -m.chat.id))
    markup.add(btn)

    mentioned_users = set()
    for k in m.entities:
        if k.type == 'mention':
            username = m.text[k.offset + 1 : k.offset + k.length].lower()
            user_id = d[username] if username in d else None
            if user_id:
                try:
                    status = bot.get_chat_member(m.chat.id, user_id).status
                    if status != 'left' and status != 'kicked':
                        mentioned_users.add((user_id, username))
                except Exception: return
    for (user_id, username) in mentioned_users:
        try:
          bot.send_message(user_id,
                           replies['main_text'].format('{} {} {}'.format(m.from_user.first_name,
                                                                         m.from_user.last_name if m.from_user.last_name else '',
                                                                         '(@{})'.format(m.from_user.username) if m.from_user.username else ''),
                                                       m.chat.title,
                                                       m.text),
                          reply_markup=markup,
                          parse_mode='HTML')
        except telebot.apihelper.ApiException as e:
          if e.result.status_code == 403 or e.result.status_code == 400:
            remove_user(username)
          else:
            print("[MAIN HANDLER EXCEPTION] [{} - {}]\n{}".format(m.from_user.username, m.from_user.id, e))

@bot.message_handler(func=lambda m: True)
def store_user(m):
  add_user(m.from_user.username, m.from_user.id)

print('Bot started:\n{}'.format(n))
bot.polling(none_stop=True)

File: tagalert.js
/****************************************************/
//   TagAlertBot (https://telegram.me/tagalertbot)  //
//   Simple notifications for mentions              //
//                                                  //
//   Author: Antonio Pitasi (@Zaphodias)            //
//   2016 - made with love                          //
/****************************************************/

const DEBUG = process.argv[2] == '--dev'

var util = require('util')
var replies = require('./replies.js')
var config = require('./config.js')
var AntiFlood = require('./antiflood.js')
var sqlite3 = require('sqlite3')
var af = new AntiFlood()
var db = new sqlite3.Database(config.dbPath)
var TelegramBot = require('node-telegram-bot-api')

var bot = new TelegramBot(config.token, {polling: {timeout: 1, interval: 1000}})

// Send a message to the admin when bot starts
bot.getMe().then((me) => {
  bot.myId = me.id
  bot.sendMessage(config.adminId, util.format(replies.booting, me.username))
})

function removeGroup(groupId) {
  db.run("DELETE FROM groups WHERE groupId=?", groupId, (err) => {
    if (err) return
    console.log("Removing group %s", groupId)
  })
}

function removeUserFromGroup(userId, groupId) {
  db.run("DELETE FROM groups WHERE userId=? AND groupId=?", userId, groupId, (err) => {
    if (err) return
    console.log("Removing @%s from group %s", userId, groupId)
  })
}

function addUser(username, userId, chatId) {
  if (!username || !userId) return

  var loweredUsername = username.toLowerCase()
  db.run("INSERT INTO users VALUES (?, ?, ?, ?)", userId, loweredUsername, + new Date(), 0, (err, res) => {
    if (err) {
      // User already in db, updating him
      db.run("UPDATE users SET username=? WHERE id=?", loweredUsername, userId, (err, res) => {})
    }
    else
      console.log("Added @%s (%s) to database", loweredUsername, userId)
  })

  if (userId !== chatId)
    db.run("INSERT INTO groups VALUES (?, ?)", chatId, userId, (err) => {})
}

function notifyUser(user, msg) {
  var notify = (userId) => {
    bot.getChatMember(msg.chat.id, userId).then((res) => {
      if (res.status == 'left' || res.status == 'kicked') return
      // User is inside in the group
      var from = util.format('%s %s %s',
        msg.from.first_name,
        msg.from.last_name ? msg.from.last_name : '',
        msg.from.username ? `(@${msg.from.username})` : ''
      )
      var btn = {inline_keyboard:[[{text: replies.retrieve}]]}
      if (msg.chat.username)
        btn.inline_keyboard[0][0].url = `telegram.me/${msg.chat.username}/${msg.message_id}`
      else
        btn.inline_keyboard[0][0].callback_data = `/retrieve_${msg.message_id}_${-msg.chat.id}`

      if (msg.photo) {
        var final_text = util.format(replies.main_caption, from, msg.chat.title, msg.caption)
        var file_id = msg.photo[0].file_id
        bot.sendPhoto(userId, file_id, {caption: final_text, reply_markup: btn})
           .then(()=>{}, ()=>{})
      }
      else {
        var final_text = util.format(replies.main_text, from, msg.chat.title, msg.text)
        bot.sendMessage(userId,
                        final_text,
                        {parse_mode: 'HTML',
                         reply_markup: btn})
           .then(()=>{}, ()=>{}) // avoid logs
      }
    })
  }

  if (user.substring) { // user is a string -> get id from db
    db.each("SELECT id FROM users WHERE username=?", user.toLowerCase(), (err, row) => {
      if (err) return
      notify(row.id)
    })
  }
  // user is a number, already the id
  else if (user.toFixed) notify(user)
}

function notifyEveryone(userId, groupId, msg) {
  db.each("SELECT userId FROM groups WHERE groupId=? AND userId!=?", groupId, userId, (err, row) => {
    if (err) return
    notifyUser(row.userId, msg)
  })
}

bot.on('callback_query', (call) => {
  if (!af.isFlooding(call.from.id)) {
    var splitted = call.data.split('_')
    if (splitted[0] === '/retrieve') {
      var messageId = splitted[1]
      var groupId = splitted[2]
      bot.sendMessage(-parseInt(groupId),
                      util.format(replies.retrieve_group, call.from.username?'@'+call.from.username:call.from.first_name),
                      {reply_to_message_id: parseInt(messageId)})
      bot.answerCallbackQuery(call.id, replies.retrieve_success, false)
    }
  }
  else bot.answerCallbackQuery(call.id, replies.flooding, false)
})

bot.onText(/\/start/, (msg) => {
  if (msg.chat.type !== 'private') return

  if (!af.isFlooding(msg.from.id)) {
    bot.sendMessage(msg.chat.id, replies.start_private, {parse_mode: 'HTML'})
  }
})

bot.onText(/^\/info$|^\/info@TagAlertBot$/gi, (msg) => {
  if (!af.isFlooding(msg.from.id)) {
    if (msg.chat.type !== 'private')
      bot.sendMessage(msg.chat.id, replies.start_group)
    else
      bot.sendMessage(msg.chat.id, replies.start_private, {parse_mode: 'HTML'})
  }
})

bot.on('message', (msg) => {
  addUser(msg.from.username, msg.from.id, msg.chat.id)

  // A user left the chat
  if (msg.left_chat_member) {
    var userId = msg.left_chat_member.id
    if (userId == bot.myId)
      removeGroup(msg.chat.id)
    else
      removeUserFromGroup(userId, msg.chat.id)
    return
  }

  if (
      (msg.chat.type !== 'group' && msg.chat.type !== 'supergroup') ||
      (msg.forward_from && msg.forward_from.id == bot.myId)
    ) return

  var toBeNotified = new Set() // avoid duplicate notifications if tagged twice

  // Text messages
  if (msg.text && msg.entities) {
    // Extract (hash)tags from message text
    var extract = (entity) => {
      return msg.text
                .substring(entity.offset + 1, entity.offset + entity.length)
                .toLowerCase()
    }

    for (var i in msg.entities) {
      var entity = msg.entities[i]

      // Tags
      if (entity.type === 'mention') {
        var username = extract(entity)
        toBeNotified.add(username)
      }

      // Hashtags
      else if (entity.type === 'hashtag') {
        var hashtag = extract(entity)
        if (hashtag === 'everyone') {
          if (!af.isFlooding(msg.from.id)) notifyEveryone(msg.from.id, msg.chat.id, msg)
        }
        else if (hashtag === 'admin') {
          bot.getChatAdministrators(msg.chat.id).then((admins) => {
            admins.forEach((admin) => { notifyUser(admin.user.id, msg) })
          })
        }
      }

      // Users without username
      else if (entity.user)
        notifyUser(entity.user.id, msg)
    }
  }

  // Images/media captions
  else if (msg.caption) {
    var matched = msg.caption.match(/@[a-z0-9]*/gi)
    for (var i in matched) {
      var username = matched[i].trim().substring(1).toLowerCase()
      toBeNotified.add(username)
    }
  }

  else return

  // helpful to check if user is tagging himself
  var isEqual = (u1, u2) => {
    if (u1 && u2) return u1.toLowerCase() === u2.toLowerCase()
    else return false
  }

  // let's really send notifications
  toBeNotified.forEach((username) => {
    // check if user is tagging himself
    if (!isEqual(msg.from.username, username) || DEBUG) {
      notifyUser(username, msg)
    }
  })
})
